### ISSUE TITLE
occurrence record weirdnesses

### ISSUE THREAD

**Issue by ayushanand18** [iobis/pyobis#45]
While working on this notebook to analyze migration patterns of species over time and plot them over a world map, I found that many records contained unexpected eventDate (screenshot attached). Should we drop these fields?
![image](https://user-images.githubusercontent.com/36472216/181775619-dd863fbf-807a-4ad6-998a-823edeeba10a.png)

**Comment by 7yl4r** [comment]
Yes, I think dropping them was the right approach. What is in the rest of the row? We may want to report these to Pieter.

**Comment by ayushanand18** [comment]
Noted. I'll update a grab here.

**Comment by ayushanand18** [comment]
Sorry for the delay, here's the grab for *Mola mola* in particular.
![image](https://user-images.githubusercontent.com/36472216/182824692-16b03988-4ea0-43db-8e6d-ca8b4b431d9f.png)

**Comment by 7yl4r** [comment]
I have been chatting with Pieter a bit, and apparently rows with no date *do* exist in OBIS. This is not worth reporting unless `date_mid` is not empty. I made [this notebook](https://colab.research.google.com/drive/1WvgPJcp82kFBuDJybH_zLTur9Cw30R3i?usp=sharing) to run some checks.

This finding looks okay because these rows do not have a `date_mid`, but let me know if you notice anything else strange.

**Comment by ayushanand18** [comment]
> I made [this notebook](https://colab.research.google.com/drive/1WvgPJcp82kFBuDJybH_zLTur9Cw30R3i?usp=sharing) to run some checks.

This notebook is really great. It is indeed clumsy to have so many columns of data that even pandas couldn't render at one go (without changing default settings.)

**Comment by ayushanand18** [comment]
I did witness another issue with OBIS data. Some of the rows of data are duplicates especially the records having MoF data. Please see the screenshot attached which contained exact same records. I couldn't apparently run the `duplicated()` function because some columns had unhashable type entries. The `date_mid` among other parameters were exact duplicates.
![image](https://user-images.githubusercontent.com/36472216/183446881-096e87c7-4231-4f61-8079-36d8d284a0fb.png)

**Comment by 7yl4r** [comment]
Can you provide the duplicated occurrenceID for some duplicated rows so I can query for the and take a look?

**Comment by ayushanand18** [comment]
> Can you provide the duplicated occurrenceID for some duplicated rows so I can query for the and take a look?

The occurrenceID for the duplicate records is `LaJollaCove_20190810_2_algae_occ3`

![image](https://user-images.githubusercontent.com/36472216/183477189-bcd93804-6416-4972-8e47-4092bd4202c3.png)

**Comment by 7yl4r** [comment]
Oops. I guess I actually need the `id` column (not `occurrenceID`) to use `occurrences.get`.

```python
pyobis.occurrences.get("LaJollaCove_20190810_2_algae_occ3")
```
> `HTTPError: 400 Client Error: Bad Request for url[...]`

I am trying to reproduce the issue [in this colab notebook](https://colab.research.google.com/drive/1KHndyH2JEMdydZqhVpxP5Y5hqbvYU2-r?usp=sharing).

Alternatively: we can use `occurrences.search` if you share the parameters used to get this data.

**Comment by albenson-usgs** [comment]
GBIF allows for searching by occurrenceID- it is only [one occurrence](https://www.gbif.org/occurrence/search?occurrence_id=LaJollaCove_20190810_2_algae_occ3&advanced=1&occurrence_status=present). There is a check that happens in the IPT to ensure each occurrenceID is unique. It's not possible to publish (via IPT at least) if there are duplicate occurrenceIDs. I think what you might be encountering is that an occurrence can have more than one measurement associated with it so if you're bringing back all the measurements you'll get repeated information for the occurrence in a flattened table because each measurement (size, weight, length, etc) will each get a row for that same occurrence.

**Comment by ayushanand18** [comment]
> Oops. I guess I actually need the UUID in the `id` column (not `occurrenceID`) to use `occurrences.get`.
> 
> ```python
> pyobis.occurrences.get("LaJollaCove_20190810_2_algae_occ3")
> ```
> 
> > `HTTPError: 400 Client Error: Bad Request for url[...]`
> 
> I am trying to reproduce the issue [in this colab notebook](https://colab.research.google.com/drive/1KHndyH2JEMdydZqhVpxP5Y5hqbvYU2-r?usp=sharing).
> 
> Alternatively: we can use `occurrences.search` if you share the parameters used to get this data.

I'm sorry for the delay, but apparently I couldn't find the record once again and therefore couldn't reproduce this error. But I found another one.

## Steps to reproduce
```python
from pyobis import occurrences as occ
data = occ.search(scientificname="Egregia menziesii", mof=True, hasextensions="MeasurementOrFacts")

# now check rows at index 26992 and 26996 they both are exactly the same
# one method is below
dict(data.loc[26992,:].T)==dict(data.loc[26996,:].T)
```
Additionally, this is the "id" parameter: 04391c9c-400a-482f-921f-023b1eb238c1

**Comment by ayushanand18** [comment]
Also, here are the details for the record
```python
{'absence': False,
 'aphiaID': 372502,
 'basisOfRecord': 'HumanObservation',
 'bathymetry': -2,
 'bibliographicCitation': nan,
 'catalogNumber': nan,
 'class': 'Phaeophyceae',
 'classid': 830,
 'collectionCode': 'NCEAS-KEEN-TS',
 'coordinateUncertaintyInMeters': '500',
 'country': nan,
 'countryCode': nan,
 'county': nan,
 'datasetID': nan,
 'datasetName': nan,
 'dataset_id': '9656db40-e4f9-4d8f-9023-38c7782efd9c',
 'date_end': 1061510400000.0,
 'date_mid': 1061510400000.0,
 'date_start': 1061510400000.0,
 'date_year': 2003.0,
 'decimalLatitude': 34.051483,
 'decimalLongitude': -119.909667,
 'depth': 4.3,
 'dropped': False,
 'eventDate': '2003-08-22',
 'eventID': nan,
 'family': 'Lessoniaceae',
 'familyid': 196198,
 'flags': [],
 'genus': 'Egregia',
 'genusid': 369610,
 'id': '04391c9c-400a-482f-921f-023b1eb238c1',
 'infrakingdom': 'Heterokonta',
 'infrakingdomid': 368898,
 'infraphylum': 'Limnista',
 'infraphylumid': 591205,
 'institutionCode': 'IMAS and University of Massachusetts',
 'kingdom': 'Chromista',
 'kingdomid': 7,
 'level': 0,
 'locality': 'SCI_FORNEY_E',
 'marine': True,
 'maximumDepthInMeters': 4.3,
 'measurementAccuracy': None,
 'measurementDeterminedBy': None,
 'measurementDeterminedDate': None,
 'measurementID': 'KEEN-TS-PISCO_subtidal-509257-indivudual-density',
 'measurementMethod': None,
 'measurementRemarks': None,
 'measurementType': 'Individual Density',
 'measurementTypeID': None,
 'measurementUnit': 'number per sq metre',
 'measurementUnitID': None,
 'measurementValue': '0',
 'measurementValueID': None,
 'minimumDepthInMeters': 4.3,
 'modified': '2021-11-30 14:51:25',
 'mof': [{'level': 0,
   'measurementAccuracy': None,
   'measurementDeterminedBy': None,
   'measurementDeterminedDate': None,
   'measurementID': 'KEEN-TS-PISCO_subtidal-509257-indivudual-density',
   'measurementMethod': None,
   'measurementRemarks': None,
   'measurementType': 'Individual Density',
   'measurementTypeID': None,
   'measurementUnit': 'number per sq metre',
   'measurementUnitID': None,
   'measurementValue': '0',
   'measurementValueID': None,
   'occurrenceID': 'KEEN-TS-PISCO_subtidal-509257'}],
 'nameAccordingTo': nan,
 'node_id': ['2a57cd59-6799-4579-955e-27c9af97aea4'],
 'occurrenceID_x': 'KEEN-TS-PISCO_subtidal-509257',
 'occurrenceID_y': 'KEEN-TS-PISCO_subtidal-509257',
 'occurrenceStatus': 'present',
 'order': 'Laminariales',
 'orderid': 845,
 'organismQuantity': 'NA',
 'organismQuantityType': 'Biomass kg wet per sq m',
 'originalScientificName': 'Egregia menziesii',
 'phylum': 'Ochrophyta',
 'phylumid': 345465,
 'recordNumber': nan,
 'recordedBy': nan,
 'samplingEffort': nan,
 'samplingProtocol': nan,
 'scientificName': 'Egregia menziesii',
 'scientificNameAuthorship': nan,
 'scientificNameID': 'urn:lsid:marinespecies.org:taxname:372502',
 'shoredistance': 405,
 'species': 'Egregia menziesii',
 'speciesid': 372502,
 'specificEpithet': nan,
 'sss': 33.44,
 'sst': 15.28,
 'stateProvince': nan,
 'subclass': 'Fucophycidae',
 'subclassid': 1304856.0,
 'subkingdom': 'Harosa',
 'subkingdomid': 582419,
 'subphylum': 'Phaeista',
 'subphylumid': 588641,
 'superclass': 'Fucistia',
 'superclassid': 588642,
 'taxonID': nan,
 'taxonRank': nan,
 'verbatimCoordinates': nan,
 'vernacularName': nan,
 'year': nan}
```

**Comment by ayushanand18** [comment]
> GBIF allows for searching by occurrenceID- it is only [one occurrence](https://www.gbif.org/occurrence/search?occurrence_id=LaJollaCove_20190810_2_algae_occ3&advanced=1&occurrence_status=present). There is a check that happens in the IPT to ensure each occurrenceID is unique within the dataset. It's not possible to publish (via IPT at least) if there are duplicate occurrenceIDs. I think what you might be encountering is that an occurrence can have more than one measurement associated with it so if you're bringing back all the measurements you'll get repeated information for the occurrence in a flattened table because each measurement (size, weight, length, etc) will each get a row for that same occurrence.

Noted, thank you for the information!

I ran an experiment on the data (which I have attached in the above comments) and found out most columns were NaNs, I could not understand why.

**Comment by albenson-usgs** [comment]
The dataset is [this one](https://obis.org/dataset/9656db40-e4f9-4d8f-9023-38c7782efd9c). I don't know why it would have repeated rows. It's being published [via an IPT](https://www.marine.csiro.au/ipt/resource?r=kelp_timeseries_keen) so it shouldn't.

**Comment by 7yl4r** [comment]
I think this may be a result of how we are handling `mof=True`. I noticed that the number of records for that query when `mof=False` is 31046 but when we pass `mof=True` the number of record rows is 208914. I think we should be getting the same number of rows on both queries?

See also:

* #61 
* [this "OBIS duplicated occurrences" google colab notebook](https://colab.research.google.com/drive/1KHndyH2JEMdydZqhVpxP5Y5hqbvYU2-r?usp=sharing)

**Comment by ayushanand18** [comment]
I think maybe [this](https://github.com/iobis/pyobis/pull/58#issuecomment-1210504409) can solve this issue.

**Comment by ayushanand18** [comment]
**Update**: Thank you so much @7yl4r and @albenson-usgs for helping out with this issue. After the fix through #64 I ran another experiment with the same query again, and it seems resolved. Here's the query.
```python
from pyobis import occurrences as occ
data = occ.search(scientificname="Egregia menziesii", mof=True, hasextensions="MeasurementOrFacts")
dict(data.loc[26992,:].T)==dict(data.loc[26996,:].T)
```
and the output:
```shell
Fetching: [████████████████████████████████████████████████████████████████████████████████████████████████████] 31046/31046
Fetched 31046 records.
False
```
